File: common/src/main/java/cc/turtl/chiselmon/api/filter/FilterDefinition.java
package cc.turtl.chiselmon.api.filter;

import cc.turtl.chiselmon.api.Priority;
import cc.turtl.chiselmon.util.format.ColorUtils;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Serializable filter definition suitable for the config format.
 *
 * <p>The condition is stored as a plain string using word-operator syntax
 * (e.g. {@code "shiny AND type=fire"}), parsed into a {@link FilterCondition}
 * tree at runtime by {@link FilterConditionParser}. This means no custom Gson
 * adapter is needed — the field serializes as a simple JSON string.
 *
 * <p><b>Migration:</b> Older saves may have a {@code tags} list instead of
 * {@code conditionString}. {@link #migrateIfNeeded()} handles this automatically.
 */
public class FilterDefinition {

    public String id;
    public String displayName;
    public int rgb;
    public Priority priority;

    /** The filter condition as a word-operator string, e.g. "shiny AND type=fire". */
    public String conditionString;

    /**
     * Legacy field — kept for migration only.
     * Gson populates this if the JSON still has a {@code "tags"} array.
     *
     * @deprecated Use {@link #conditionString} instead.
     */
    @Deprecated
    public List<String> tags;

    public FilterDefinition(String id, String displayName, int rgb, Priority priority,
                            String conditionString) {
        this.id = id;
        this.displayName = displayName;
        this.rgb = rgb;
        this.priority = priority;
        this.conditionString = conditionString;
    }

    /**
     * Migrates from the old {@code tags} list to a {@code conditionString}.
     * No-op if {@code conditionString} is already set.
     */
    public void migrateIfNeeded() {
        if (conditionString != null) return;

        if (tags != null && !tags.isEmpty()) {
            conditionString = String.join(" AND ", tags);
        } else {
            conditionString = "";
        }
        tags = null;
    }

    // -------------------------------------------------------------------------
    // Default filters
    // -------------------------------------------------------------------------

    public static class DefaultFilters {

        public static final FilterDefinition LEGENDARIES = new FilterDefinition(
                "legendaries", "Legendary Pokemon",
                ColorUtils.MAGENTA.getRGB(), Priority.HIGHEST,
                "legendary"
        );

        public static final FilterDefinition SHINIES = new FilterDefinition(
                "shinies", "Shiny Pokemon",
                ColorUtils.GOLD.getRGB(), Priority.HIGH,
                "shiny"
        );

        public static final FilterDefinition EXTREME_SIZES = new FilterDefinition(
                "extreme_sizes", "Extreme Size Pokemon",
                ColorUtils.TEAL.getRGB(), Priority.NORMAL,
                "extreme_size"
        );

        private static final List<FilterDefinition> ALL = List.of(LEGENDARIES, SHINIES, EXTREME_SIZES);

        public static Map<String, FilterDefinition> all() {
            return ALL.stream().collect(Collectors.toUnmodifiableMap(f -> f.id, f -> f));
        }
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/filter/FilterDefinition.kt
package cc.turtl.chiselmon.core.api.filter

import cc.turtl.chiselmon.util.format.ColorUtils
import cc.turtl.turtlshell.api.core.Priority

/**
 * Serializable filter definition suitable for the config format.
 *
 * The condition is stored as a plain string using word-operator syntax
 * (e.g. "shiny AND type=fire"), parsed into a [FilterCondition] tree at
 * runtime by [FilterConditionParser]. No custom Gson adapter is needed --
 * the field serializes as a simple JSON string.
 */
class FilterDefinition(
    @JvmField var id: String,
    @JvmField var displayName: String,
    @JvmField var rgb: Int,
    @JvmField var priority: Priority,
    /** The filter condition as a word-operator string, e.g. "shiny AND type=fire". */
    @JvmField var conditionString: String
) {
    /**
     * Legacy field kept for migration only.
     * Gson populates this if the JSON still has a "tags" array.
     */
    @Deprecated("Use conditionString instead")
    @JvmField
    var tags: List<String>? = null

    // -------------------------------------------------------------------------
    // Default filters
    // -------------------------------------------------------------------------

    object DefaultFilters {

        @JvmField
        val LEGENDARIES = FilterDefinition(
            "legendaries", "Legendary Pokemon",
            ColorUtils.MAGENTA.rgb, Priority.HIGHEST,
            "legendary"
        )

        @JvmField
        val SHINIES = FilterDefinition(
            "shinies", "Shiny Pokemon",
            ColorUtils.GOLD.rgb, Priority.HIGH,
            "shiny"
        )

        @JvmField
        val EXTREME_SIZES = FilterDefinition(
            "extreme_sizes", "Extreme Size Pokemon",
            ColorUtils.TEAL.rgb, Priority.NORMAL,
            "extreme_size"
        )

        private val ALL = listOf(LEGENDARIES, SHINIES, EXTREME_SIZES)

        @JvmStatic
        fun all(): Map<String, FilterDefinition> = ALL.associateBy { it.id }
    }
}